File: src/dj.rb
require_relative 'database'
db = Database.new

class DJ
  def self.playlist(tracks, length)
    tracks.each { |track| track["weight"] = determine_weight(track) }
    tracks.select { |track| 
      track["weight"] != 0 
    }
    self.mix_tracks(tracks)[...length]
  end

  def self.mix_tracks(tracks)
    totalWeight = tracks.reduce(0) { |sum, track| sum + track["weight"] }
    tracks.each do |track| 
      track["randomizedPriority"] = rand ** (totalWeight / track["weight"])
    end
    return tracks.sort {|a, b| b["randomizedPriority"] <=> a["randomizedPriority"]}
  end

  def self.determine_weight(track)
    return 0 if track["deleted"] == 1

    factors = []

    if track["length"] >= 480 
      factors.push 480 / track["length"]  
    end

    if track["upvotes"] > 0
      factors.push track["upvotes"]
    else
      factors.push 0.8 ** track["times_listened"]
    end

    return factors.reduce(1, :*).to_f 
  end
end

File: src/dj.js
export class DJ {
  static playlist(tracks) {
    tracks.forEach(track => {
      track.weight = this.determineWeight(track);
    });
    const filteredTracks = tracks.filter(track => track.weight !== 0);
    return this.mixTracks(filteredTracks);
  }

  static mixTracks(tracks) {
    const totalWeight = tracks.reduce((sum, track) => sum + track.weight, 0);
    tracks.forEach(track => {
      track.randomizedPriority = Math.pow(Math.random(), totalWeight / track.weight);
    });
    return tracks.sort((a, b) => b.randomizedPriority - a.randomizedPriority);
  }

  static determineWeight(track) {
    let factors = [];

    // todo: check that this is seconds
    if (track.length >= 480) {
      factors.push(480.0 / track.length);
    }

    if (track.upvotes > 0) {
      factors.push(track.upvotes);
    } else {
      factors.push(Math.pow(0.8, track.times_listened));
    }

    return factors.reduce((acc, val) => acc * val, 1);
  }
}

